src/main/java/bio/terra/service/filedata/google/gcs/GcsProjectFactory.java

```diff
@@ -1,34 +1,104 @@
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
 
 @Component
 public class GcsProjectFactory {
-  private static final ConcurrentHashMap<String, GcsProject> gcsProjectCache =
-      new ConcurrentHashMap<>();
+
+  record ProjectUserIdentifier(String projectId, String userToImpersonate) {}
+
+  private static final Map<ProjectUserIdentifier, GcsProject> gcsProjectCache =
+      Collections.synchronizedMap(
+          new PassiveExpiringMap<>(GcsProject.TOKEN_LENGTH.toSeconds(), TimeUnit.SECONDS));
+
+  private static final int PROJECT_RESOURCES_CACHE_SIZE = 500;
+  private static final Map<String, Optional<GoogleProjectResource>> projectResources =
+      Collections.synchronizedMap(new LRUMap<>(PROJECT_RESOURCES_CACHE_SIZE));
 
   private final GcsConfiguration gcsConfiguration;
+  private final GoogleResourceDao googleResourceDao;
 
   @Autowired
-  public GcsProjectFactory(GcsConfiguration gcsConfiguration) {
+  public GcsProjectFactory(GcsConfiguration gcsConfiguration, GoogleResourceDao googleResourceDao) {
     this.gcsConfiguration = gcsConfiguration;
+    this.googleResourceDao = googleResourceDao;
   }
 
+  /**
+   * Get a GcsProject object for the specified project authorized with a project-specific service
+   * account
+   *
+   * @param projectId The project ID to get a GcsProject object for
+   * @return GcsProject metadata object
+   */
   public GcsProject get(String projectId) {
-    gcsProjectCache.computeIfAbsent(
-        projectId,
+    return get(projectId, false);
+  }
+
+  /**
+   * Get a GcsProject object for the specified project authorized with either a project-specific
+   * service account or the TDR service account
+   *
+   * @param projectId The project ID to get a GcsProject object for
+   * @param asMainServiceAccount If true, authorize using the main TDR service account
+   * @return GcsProject metadata object
+   */
+  public GcsProject get(String projectId, boolean asMainServiceAccount) {
+    GoogleProjectResource projectResource;
+
+    ProjectUserIdentifier projectUserIdentifier;
+    if (asMainServiceAccount || (projectResource = retrieveProjectResource(projectId)) == null) {
+      projectUserIdentifier = new ProjectUserIdentifier(projectId, null);
+    } else {
+      projectUserIdentifier =
+          new ProjectUserIdentifier(projectId, projectResource.getServiceAccount());
+    }
+
+    return gcsProjectCache.computeIfAbsent(
+        projectUserIdentifier,
         p ->
             new GcsProject(
                 p,
                 gcsConfiguration.getConnectTimeoutSeconds(),
                 gcsConfiguration.getReadTimeoutSeconds()));
-    return gcsProjectCache.get(projectId);
   }
 
+  private GoogleProjectResource retrieveProjectResource(String projectId) {
+    return projectResources
+        .computeIfAbsent(projectId, googleResourceDao::retrieveProjectByGoogleProjectIdMaybe)
```

Comment:
It looks like this is caching the contents of our database. Is the database access a source of performance issues? How is the cache invalidated if the database content is changed?